### PR TITLE
refactor: merge catalog provider & schema provider into catalog manager

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Remove `CatalogProvider` and `SchemaProvider`. Move all the necessary methods into `CatalogManager`. 

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by waynexia** [comment]
All UT pass now. I'm going to resolve conflict and check other tests later. This is generally ready for review

**Review by evenyag — APPROVED** [review]
Since this PR has passed all test cases, I choose to believe @waynexia

File: src/catalog/src/local/manager.rs [hunk 1]
```diff
@@ -116,26 +117,44 @@ impl LocalCatalogManager {
...
+            .register_catalog_sync(DEFAULT_CATALOG_NAME.to_string())?;
+        self.catalogs.register_schema_sync(RegisterSchemaRequest {
+            catalog: DEFAULT_CATALOG_NAME.to_string(),
+            schema: DEFAULT_SCHEMA_NAME.to_string(),
+        })?;
 
         // Add numbers table for test
-        let table = Arc::new(NumbersTable::default());
-        default_schema.register_table_sync("numbers".to_string(), table)?;
+        let numbers_table = Arc::new(NumbersTable::default());
```
Comment (v0y4g3r) [inline src/catalog/src/local/manager.rs:145]: Can we remove numbers table in this PR?

File: src/catalog/src/local/memory.rs [hunk 2]
```diff
@@ -50,13 +48,15 @@ impl Default for MemoryCatalogManager {
             table_id: AtomicU32::new(MIN_USER_TABLE_ID),
             catalogs: Default::default(),
         };
-        let default_catalog = Arc::new(MemoryCatalogProvider::new());
+
+        let mut catalog = HashMap::new();
```
Comment (v0y4g3r) [inline src/catalog/src/local/memory.rs:52]:
```suggestion
        let mut catalog = HashMap::with_capacity(1);
```

File: src/catalog/src/local/memory.rs [hunk 3]
```diff
@@ -76,80 +76,75 @@ impl CatalogManager for MemoryCatalogManager {
...
+            .fail()?;
+        }
+        if schema.contains_key(&request.new_table_name) {
+            return TableExistsSnafu {
+                table: &request.new_table_name,
+            }
+            .fail();
+        }
+
+        let table = schema.remove(&request.table_name).unwrap();
```
Comment (v0y4g3r) [inline src/catalog/src/local/memory.rs:117]: Is it safe to unwrap here?

File: src/catalog/src/local/memory.rs [hunk 4]
```diff
@@ -76,80 +76,75 @@ impl CatalogManager for MemoryCatalogManager {
...
-            .map(|v| v.is_some())
+        let result = schema.remove(&request.table_name);
+        if result.is_some() {
+            decrement_gauge!(
+                crate::metrics::METRIC_CATALOG_MANAGER_TABLE_COUNT,
+                1.0,
+                &[crate::metrics::db_label(&request.catalog, &request.schema)],
+            );
+        }
+        Ok(result.is_some())
```
Comment (v0y4g3r) [inline src/catalog/src/local/memory.rs:143]: As we've discussed last week, can we simply change the return value type from `Result<bool>` to `Result<()>`?

File: src/frontend/src/catalog.rs [hunk 5]
```diff
@@ -409,20 +298,71 @@ impl SchemaProvider for FrontendSchemaProvider {
...
+    async fn table(
+        &self,
+        catalog: &str,
+        schema: &str,
+        table_name: &str,
+    ) -> CatalogResult<Option<TableRef>> {
+        if catalog == DEFAULT_CATALOG_NAME
+            && schema == DEFAULT_SCHEMA_NAME
+            && table_name == "numbers"
         {
```
Comment (v0y4g3r) [inline src/frontend/src/catalog.rs:342]: I think we should remove numbers table, again.

File: src/catalog/src/local/memory.rs (hunk 3 above)
Comment (waynexia) [reply src/catalog/src/local/memory.rs:117]: Yes, this is checked before via `contains_key()`.

File: src/catalog/src/local/manager.rs (hunk 1 above)
Comment (waynexia) [reply src/catalog/src/local/manager.rs:145]: Agree, we need to reconsider it. Filed https://github.com/GreptimeTeam/greptimedb/issues/1827

**Review by v0y4g3r — APPROVED** [review]
a small step for catalog, a giant leap for greptimedb

**Pull request merged**